e2e/ethereum_test.go

```diff
@@ -0,0 +1,189 @@
+	//deployPK string, chainID *big.Int, url string, relayers int, initialRelayerThreshold *big.Int, minCount uint8
+	contracts, err := ethereum.DeployContracts(
+		hexutil.Encode(AliceEthKp.Encode())[2:],
+		big.NewInt(int64(id)),
+		TestEthEndpoint,
+		3,             // Num relayers
+		big.NewInt(2), // Relayer threshold
+		0,             // unused?
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fmt.Println("====== Contracts =======")
+	fmt.Printf("Bridge: %s\n", contracts.BridgeAddress.Hex())
+	fmt.Printf("Erc20Handler: %s\n", contracts.ERC20HandlerAddress.Hex())
+	fmt.Printf("(Generic) Centrifuge Handler: %s\n", contracts.CentrifugeHandlerAddress.Hex())
+	fmt.Println("========================")
+	return contracts
+}
+
+func createEthClient(t *testing.T) (*ethclient.Client, *bind.TransactOpts) {
+	ctx := context.Background()
+	rpcClient, err := rpc.DialWebsocket(ctx, TestEthEndpoint, "/ws")
+	if err != nil {
+		t.Fatal(err)
+	}
+	client := ethclient.NewClient(rpcClient)
+
+	nonce, err := client.PendingNonceAt(ctx, CharlieEthAddr)
+	if err != nil {
+		t.Fatal(err)
+	}
+	opts := bind.NewKeyedTransactor(CharlieEthKp.PrivateKey())
+	opts.Nonce = big.NewInt(int64(nonce - 1))        // -1 since we always increment before calling
+	opts.Value = big.NewInt(0)                       // in wei
+	opts.GasLimit = uint64(ethereum.DefaultGasLimit) // in units
+	opts.GasPrice = big.NewInt(ethereum.DefaultGasPrice)
+	opts.Context = ctx
+
+	return client, opts
+}
+
+// deployMintApproveErc20 funds the account with a newly created erc20, to prepare for initiating a transfer
+func deployMintApproveErc20(t *testing.T, client *ethclient.Client, opts *bind.TransactOpts, erc20Handler common.Address) common.Address {
+	// Deploy
+	opts.Nonce = opts.Nonce.Add(opts.Nonce, big.NewInt(1))
+	erc20Addr, _, erc20Instance, err := erc20Mintable.DeployERC20Mintable(opts, client)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Mint
+	opts.Nonce = opts.Nonce.Add(opts.Nonce, big.NewInt(1))
+	_, err = erc20Instance.Mint(opts, opts.From, big.NewInt(99))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Approve
+	fmt.Printf("Approving: %s\n", erc20Handler.Hex())
+	opts.Nonce = opts.Nonce.Add(opts.Nonce, big.NewInt(1))
+	_, err = erc20Instance.Approve(opts, erc20Handler, big.NewInt(99))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return erc20Addr
+}
+
+// deployAndFundEr20 sets up a new erc20 contract and funds the bridge/handler
+func deployAndFundErc20(t *testing.T, client *ethclient.Client, opts *bind.TransactOpts, erc20Handler common.Address) common.Address { //nolint:unused,deadcode
+	// Deploy
+	opts.Nonce = opts.Nonce.Add(opts.Nonce, big.NewInt(1))
+	erc20Addr, _, erc20Instance, err := erc20Mintable.DeployERC20Mintable(opts, client)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Mint to handler
+	opts.Nonce = opts.Nonce.Add(opts.Nonce, big.NewInt(1))
+	_, err = erc20Instance.Mint(opts, erc20Addr, big.NewInt(99))
```

Comment:
Why do you fund the bridge? Shouldn't you fund a user depositing the tokens?

Reply:
This is for the inverse operation, the bridge releasing tokens to a user as the result of a bridge transfer (test not yet implemented)